### ISSUE TITLE
alpha feature gate: LegacyServiceAccountTokenCleanUp failed to get configmap kube-apiserver-legacy-service-account-token-tracking

### ISSUE THREAD

**Issue by rxwang662001** [kubernetes/kubernetes#120298]
### What happened?

installed k8s 1.28.1, 
and enable controller manager's feature gate LegacyServiceAccountTokenCleanUp as below:
   - --feature-gates=LegacyServiceAccountTokenCleanUp=true
   - --legacy-service-account-token-clean-up-period=1h

Then controller-manager shall be restarted, but in the log it shows error
```
I0831 08:04:39.997150       1 legacy_serviceaccount_token_cleaner.go:101] "Starting legacy service account token cleaner controller"
I0831 08:04:39.997165       1 shared_informer.go:311] Waiting for caches to sync for legacy-service-account-token-cleaner
I0831 08:04:40.297417       1 shared_informer.go:318] Caches are synced for legacy-service-account-token-cleaner
E0831 08:04:40.304939       1 legacy_serviceaccount_token_cleaner.go:119] "Getting lastest possible tracked_since time" err="configmaps \"kube-apiserver-legacy-service-account-token-tracking\" is forbidden: User \"system:serviceaccount:kube-system:legacy-service-account-token-cleaner\" cannot get resource \"configmaps\" in API group \"\" in the namespace \"kube-system\""
```
If manually binding a clusterrole to the serviceaccount:kube-system:legacy-service-account-token-cleaner and restart controller-manager, there is no such error then.

Please let me know whether I missed anything or this is an issue.

and one more, 
since the "kubernetes.io/legacy-token-last-used:" is recorded as dateOnly base,
so the clean shall take effect at least 1 day later？
labels:
    kubernetes.io/legacy-token-last-used: "2023-08-31"

and since DefaultCleanerSyncInterval = 24 * time.Hour, the cleaner works every 24 hours?
looks no way to change the default now by the user self?

### What did you expect to happen?

no controller-manager log error when enabling feature gate LegacyServiceAccountTokenCleanUp

### How can we reproduce it (as minimally and precisely as possible)?

just install k8s 1.28.1 and enable the feature gate LegacyServiceAccountTokenCleanUp

### Kubernetes version

<details>

```console
$ kubectl version
 kubectl version
Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.1

```

</details>

### Cloud provider

<details>
private installed k8s
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by yt2985** [comment]
Hi @rxwang662001, it looks there is something wrong with the cleaner policy and thus it could not get the required roles, though I indeed added the required rules: https://github.com/yt2985/kubernetes/blob/595a57940a45f25bad4621548ee990c81779c12e/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go#L474 Unfortunately, in alpha, I did not have a time to test the policy. I will check that and create a fix if something is wrong. Thank you for reporting that!

**Comment by yt2985** [comment]
Hi @rxwang662001, I tested by creating a cluster with v1.28 and enabling the feature. But I could not reproduce the error 
```
"Getting lastest possible tracked_since time" err="configmaps \"kube-apiserver-legacy-service-account-token-tracking\" is forbidden: User \"system:serviceaccount:kube-system:legacy-service-account-token-cleaner\" cannot get resource \"configmaps\" in API group \"\" in the namespace \"kube-system\""
```

I also observed that the ClusterRole and ClusterRoleBinding for `system:controller:legacy-service-account-token-cleaner` were created using `kubectl get clusterrole` and `kubectl get clusterrolebinding` command. 

I further checked by running `kubectl auth can-i get configmaps/kube-apiserver-legacy-service-account-token-tracking --namespace kube-system --as system:serviceaccount:kube-system:legacy-service-account-token-cleaner`. The result is `yes`. 

It's strange that you observed the reported error. Would you please try it again and see whether the error insists?

**Comment by yt2985** [comment]
hmmm, please neglect my previous reply. I now reproduced the same error in another cluster (which I only enabled legacy-service-account-token-cleaner feature, previously I tested in the alpha cluster with all the alpha features enabled). I will keep investigating the problem reason.

**Comment by yt2985** [comment]
> hmmm, please neglect my previous reply. I now reproduced the same error in another cluster (which I only enabled legacy-service-account-token-cleaner feature, previously I tested in the alpha cluster with all the alpha features enabled). I will keep investigating the problem reason.

I have a guess. For the time that I reproduced the error, I firstly created a cluster with v1.28 (without enabling the feature), and then manually changed the kube-controller-manager yaml file by adding the flags. Since the clusterroles and clusterrolebinds are created during the bootstrapping (https://sourcegraph.com/github.com/kubernetes/kubernetes@133eff3df4951a204a1acecd3a25afa3e3967c32/-/blob/pkg/registry/rbac/rest/storage_rbac.go?L133#:~:text=132-,133,-134), it is possibly that these clusterroles and clusterrolebings will not be created again after a later feature enabling. For the alpha cluster, the feature is enabled from the initial time, so the clusterroles and clusterrolebings are created.

**Comment by liggitt** [comment]
yes, the feature gate should be consistently enabled in kube-apiserver and kube-controller-manager

**Comment by rxwang662001** [comment]
Thanks @liggitt  @yt2985 
Yes, after enabling the feature gate in apiserver also, the error is gone.
After thinking it back, generally there is no place to clearly state which components need to be set when enabling one feature gate (sure, we'd like to expect there is such info in somewhere like k8s feature gate page）, so I shall check each component like apiserver/controller manager/scheduller/kubelet/kubeproxy's configuration to see whether there is feature gate for it, but for this one I remembered in k8s doc either release note or KEP page there is statement it's for controller manager, that's why I set it on controller manager only, which may be needed to change if there does have such.
Anyway, it's working now, I shall close this ticket, thanks.
